build-support/bin/generate_docs.py

```diff
@@ -0,0 +1,301 @@
+        )
+        self._category_id = None  # Fetched lazily.
+
+        # Load the data.
+        if self._args.input is None:
+            json_str = sys.stdin.read()
+        else:
+            with open(self._args.input, "r") as fp:
+                json_str = fp.read()
+        self._help_info = self.process_input(json_str, self._args.sync)
+
+    @staticmethod
+    def process_input(json_str: str, sync: bool):
+        """Process the input, to make it easier to work with in the mustache template."""
+
+        help_info = json.loads(json_str)
+        scope_to_help_info = help_info["scope_to_help_info"]
+
+        # Process the list of consumed_scopes into a comma-separated list, and add it to the option
+        # info for the goal's scope, to make it easy to render in the goal's options page.
+
+        # docsite pages link to the slug, local pages to the .md source.
+        link_suffix = "-reference" if sync else ".md"
+        for goal, goal_info in help_info["name_to_goal_info"].items():
+            consumed_scopes = sorted(goal_info["consumed_scopes"])
+            linked_consumed_scopes = [f"[{cs}]({cs}{link_suffix})" for cs in consumed_scopes if cs]
+            comma_separated_consumed_scopes = ", ".join(linked_consumed_scopes)
+            scope_to_help_info[goal][
+                "comma_separated_consumed_scopes"
+            ] = comma_separated_consumed_scopes
+
+        # Process the option data.
+
+        def munge_option(option_data):
+            # Munge the default str so we can display it nicely when it's multiline, while
+            # still displaying it inline if it's not.
+            default_str = option_data["default_str"]
+            escaped_default_str = html.escape(default_str, quote=False)
+            if "\n" in default_str:
+                option_data["marked_up_default"] = f"<pre>{escaped_default_str}</pre>"
+            else:
+                option_data["marked_up_default"] = f"<code>{escaped_default_str}</code>"
+
+        for shi in scope_to_help_info.values():
+            for opt in shi["basic"]:
+                munge_option(opt)
+            for opt in shi["advanced"]:
+                munge_option(opt)
+
+        return help_info
+
+    @property
+    def category_id(self):
+        """The id of the "Reference" category on the docsite."""
+        if self._category_id is None:
+            self._category_id = self._get_id("categories/reference")
+        return self._category_id
+
+    def _render_body(self, scope_help_info):
+        """Renders the body of a single options help page."""
+        return self._renderer.render("{{> scoped_options}}", scope_help_info)
+
+    def _access_readme_api(self, url_suffix: str, method: str, payload: str) -> Dict:
+        """Sends requests to the readme.io API."""
+        url = f"https://dash.readme.io/api/v1/{url_suffix}"
+        version = "v2.0"  # TODO: Don't hardcode this.
```

Comment:
Is this what will allow us to update 2.0 vs. 2.1 and so on?

Reply:
Yes, and we can ~figure out what to update based on the current pants version. But I decided to punt for now.